core/types/block.go

```diff
@@ -454,30 +455,15 @@ func (b *Block) WithSeal(header *Header) *Block {
 }
 
 // WithBody returns a copy of the block with the given transaction and uncle contents.
-func (b *Block) WithBody(transactions []*Transaction, uncles []*Header) *Block {
-	block := &Block{
-		header:       b.header,
-		transactions: make([]*Transaction, len(transactions)),
-		uncles:       make([]*Header, len(uncles)),
-		withdrawals:  b.withdrawals,
-	}
-	copy(block.transactions, transactions)
-	for i := range uncles {
-		block.uncles[i] = CopyHeader(uncles[i])
-	}
-	return block
-}
-
-// WithWithdrawals returns a copy of the block containing the given withdrawals.
-func (b *Block) WithWithdrawals(withdrawals []*Withdrawal) *Block {
+func (b *Block) WithBody(body Body) *Block {
```

Comment:
Method doc should expand a bit. E.g. 
```
// WithBody returns a new block, consisting of the orginal header (not copied) along
// with the body contents (which are deep-copied).
// OBS: The header-fields based on body derivatives (e..g txhash) are _not_ updated, and will
// mismatch if the body does not match the header.
```
It's a bit funny / curious that this method avoids copying the original header, but conscientiously copies the incoming body.

Reply:
So the idea here is, header fields inside `types.Block` can't leak out because we copy their values in block accessors. We do need to copy the fields in the passed `Body` here because a later modification of the input data (e.g. uncle header values) could mess up the internal caches in `Block`.